File: packages/react-native/src/private/specs/modules/NativeExceptionsManager.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 * @format
 */

import type {TurboModule} from '../../../../Libraries/TurboModule/RCTExport';

import * as TurboModuleRegistry from '../../../../Libraries/TurboModule/TurboModuleRegistry';

const Platform = require('../../../../Libraries/Utilities/Platform');

export type StackFrame = {|
  column: ?number,
  file: ?string,
  lineNumber: ?number,
  methodName: string,
  collapse?: boolean,
|};
export type ExceptionData = {
  message: string,
  originalMessage: ?string,
  name: ?string,
  componentStack: ?string,
  stack: Array<StackFrame>,
  id: number,
  isFatal: boolean,
  // flowlint-next-line unclear-type:off
  extraData?: Object,
  ...
};
export interface Spec extends TurboModule {
  // Deprecated: Use `reportException`
  +reportFatalException: (
    message: string,
    stack: Array<StackFrame>,
    exceptionId: number,
  ) => void;
  // Deprecated: Use `reportException`
  +reportSoftException: (
    message: string,
    stack: Array<StackFrame>,
    exceptionId: number,
  ) => void;
  +reportException?: (data: ExceptionData) => void;
  // TODO(T53311281): This is a noop on iOS now. Implement it.
  +dismissRedbox?: () => void;
}

const NativeModule =
  TurboModuleRegistry.getEnforcing<Spec>('ExceptionsManager');

const ExceptionsManager = {
  reportFatalException(
    message: string,
    stack: Array<StackFrame>,
    exceptionId: number,
  ) {
    NativeModule.reportFatalException(message, stack, exceptionId);
  },
  reportSoftException(
    message: string,
    stack: Array<StackFrame>,
    exceptionId: number,
  ) {
    NativeModule.reportSoftException(message, stack, exceptionId);
  },
  dismissRedbox(): void {
    if (Platform.OS !== 'ios' && NativeModule.dismissRedbox) {
      // TODO(T53311281): This is a noop on iOS now. Implement it.
      NativeModule.dismissRedbox();
    }
  },
  reportException(data: ExceptionData): void {
    if (NativeModule.reportException) {
      NativeModule.reportException(data);
      return;
    }
    if (data.isFatal) {
      ExceptionsManager.reportFatalException(data.message, data.stack, data.id);
    } else {
      ExceptionsManager.reportSoftException(data.message, data.stack, data.id);
    }
  },
};

export default ExceptionsManager;

File: packages/react-native/ReactCxxPlatform/react/logging/NativeExceptionsManager.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "NativeExceptionsManager.h"

#include <utility>

namespace facebook::react {

void NativeExceptionsManager::reportFatalException(
    jsi::Runtime& rt,
    std::string message,
    std::vector<StackFrame> stack,
    int32_t exceptionId) {
  reportException(
      rt,
      ExceptionData{
          .message = std::move(message),
          .stack = std::move(stack),
          .id = exceptionId,
          .isFatal = true});
}

void NativeExceptionsManager::reportSoftException(
    jsi::Runtime& rt,
    std::string message,
    std::vector<StackFrame> stack,
    int32_t exceptionId) {
  reportException(
      rt,
      ExceptionData{
          .message = std::move(message),
          .stack = std::move(stack),
          .id = exceptionId,
          .isFatal = false});
}

void NativeExceptionsManager::reportException(
    jsi::Runtime& runtime,
    const ExceptionData& data) {
  if (onJsError_) {
    std::vector<JsErrorHandler::ProcessedError::StackFrame> frames;
    frames.reserve(data.stack.size());
    for (const auto& frame : data.stack) {
      frames.push_back(JsErrorHandler::ProcessedError::StackFrame{
          .file = frame.file,
          .methodName = frame.methodName,
          .lineNumber = frame.lineNumber,
          .column = frame.column});
    }
    JsErrorHandler::ProcessedError processedError{
        .message = data.message,
        .originalMessage = std::nullopt,
        .name = std::nullopt,
        .componentStack = std::nullopt,
        .stack = frames,
        .id = data.id,
        .isFatal = data.isFatal,
        .extraData = jsi::Object(runtime),
    };
    onJsError_(runtime, processedError);
  }
}

} // namespace facebook::react
